// @decentralized-identity/hub-common-js/dist/src/responses/HubErrorCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * List of error codes potentially returned by an Identity Hub.
 */
var HubErrorCode;
(function (HubErrorCode) {
    /** Indicates that the attempted authentication method was invalid or expired. */
    HubErrorCode["AuthenticationFailed"] = "authentication_failed";
    /** Indicates that the request issued by the client was invalid. */
    HubErrorCode["BadRequest"] = "bad_request";
    /** Indicates that the requested entity was not found. */
    HubErrorCode["NotFound"] = "not_found";
    /** Indicates that the requested interface/method is not yet implemented. */
    HubErrorCode["NotImplemented"] = "not_implemented";
    /** Indicates that the client lacks necessary permissions to complete the request. */
    HubErrorCode["PermissionsRequired"] = "permissions_required";
    /** Indicates that the Hub service is temporarialy unavailable. */
    HubErrorCode["TemporarilyUnavailable"] = "temporarily_unavailable";
    /** Indicates that the client has made too many requests recently and should back off attempts. */
    HubErrorCode["TooManyRequests"] = "too_many_requests";
    /** Indicates that an internal error occurred inside the Hub. */
    HubErrorCode["ServerError"] = "server_error";
    /** Indicates that the Hub service is temporarialy unavailable. */
    HubErrorCode["ServiceUnavailable"] = "service_unavailable";
})(HubErrorCode || (HubErrorCode = {}));
exports.default = HubErrorCode;
