// @decentralized-identity/did-auth-jose/dist/lib/crypto/ec/EcPublicKey.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const PublicKey_1 = __importStar(require("../../security/PublicKey"));
/**
 * Represents an Elliptic Curve public key
 * @class
 * @extends PublicKey
 */
class EcPublicKey extends PublicKey_1.default {
    /**
     * An Elliptic Curve JWK
     * @param keyData The IDidDocumentPublicKey containing the elliptic curve public key parameters.
     */
    constructor(keyData) {
        super();
        this.kty = PublicKey_1.RecommendedKeyType.Ec;
        this.kid = keyData.id;
        const data = keyData;
        if ('publicKeyJwk' in data) {
            const jwk = data.publicKeyJwk;
            if (!keyData.id.endsWith(jwk.kid)) {
                throw new Error('JWK kid does not match Did publickey id.');
            }
            if (!jwk.crv || !jwk.x || !jwk.y) {
                throw new Error('JWK missing required parameters.');
            }
            this.crv = jwk.crv;
            this.x = jwk.x;
            this.y = jwk.y;
            this.key_ops = jwk.key_ops;
            this.use = this.use;
        }
        else {
            throw new Error('Cannot parse Elliptic Curve key.');
        }
    }
}
exports.default = EcPublicKey;
