// @decentralized-identity/hub-common-js/dist/src/index.js
"use strict";
/**
 * Defines package exports for @decentralized-identity/hub-common-js.
 */
Object.defineProperty(exports, "__esModule", { value: true });
// Commits
var CommitOperation_1 = require("./commits/CommitOperation");
exports.CommitOperation = CommitOperation_1.default;
var TestCommit_1 = require("./commits/TestCommit");
// Commit Strategies
var CommitStrategy_1 = require("./commitStrategy/CommitStrategy");
exports.CommitStrategy = CommitStrategy_1.default;
var BasicCommitStrategy_1 = require("./commitStrategy/BasicCommitStrategy");
exports.BasicCommitStrategy = BasicCommitStrategy_1.default;
exports.BASIC_COMMIT_STRATEGY = BasicCommitStrategy_1.BASIC_COMMIT_STRATEGY;
var LastWriterWinsCommitStrategy_1 = require("./commitStrategy/LastWriterWinsCommitStrategy");
exports.LastWriterWinsCommitStrategy = LastWriterWinsCommitStrategy_1.default;
exports.LAST_WRITER_WINS_COMMIT_STRATEGY = LastWriterWinsCommitStrategy_1.LAST_WRITER_WINS_COMMIT_STRATEGY;
var HubErrorCode_1 = require("./responses/HubErrorCode");
exports.HubErrorCode = HubErrorCode_1.default;
// Unit Test utilities
exports.unitTestExports = {
    TestCommit: TestCommit_1.default,
};
